stubs/SQLAlchemy/sqlalchemy/engine/base.pyi

```diff
@@ -1,61 +1,97 @@
-from typing import Any
+from _typeshed import Self
+from abc import abstractmethod
+from collections.abc import Mapping
+from types import TracebackType
+from typing import Any, Callable, TypeVar, overload
 
-from .. import log
-from .interfaces import Connectable as Connectable, ExceptionContext
+from ..dbapi import DBAPIConnection
+from ..log import Identified, echo_property
+from ..pool import Pool
+from ..sql.compiler import Compiled
+from ..sql.ddl import DDLElement
+from ..sql.elements import ClauseElement
+from ..sql.functions import FunctionElement
+from ..sql.schema import DefaultGenerator
+from .cursor import CursorResult
+from .interfaces import Connectable as Connectable, Dialect, ExceptionContext
 from .util import TransactionalContext
 
+_T = TypeVar("_T")
+
 class Connection(Connectable):
-    engine: Any
-    dialect: Any
+    engine: Engine
+    dialect: Dialect
     should_close_with_result: bool
     dispatch: Any
     def __init__(
         self,
-        engine,
-        connection: Any | None = ...,
+        engine: Engine,
+        connection: DBAPIConnection | None = ...,
         close_with_result: bool = ...,
         _branch_from: Any | None = ...,
         _execution_options: Any | None = ...,
         _dispatch: Any | None = ...,
         _has_events: Any | None = ...,
         _allow_revalidate: bool = ...,
     ) -> None: ...
-    def schema_for_object(self, obj): ...
-    def __enter__(self): ...
-    def __exit__(self, type_, value, traceback) -> None: ...
+    def schema_for_object(self, obj) -> str: ...
+    def __enter__(self: Self) -> Self: ...
+    def __exit__(
+        self, type_: type[BaseException] | None, value: BaseException | None, traceback: TracebackType | None
+    ) -> None: ...
     def execution_options(self, **opt): ...
     def get_execution_options(self): ...
     @property
-    def closed(self): ...
+    def closed(self) -> bool: ...
     @property
-    def invalidated(self): ...
+    def invalidated(self) -> bool: ...
     @property
-    def connection(self): ...
+    def connection(self) -> DBAPIConnection: ...
     def get_isolation_level(self): ...
     @property
     def default_isolation_level(self): ...
     @property
     def info(self): ...
     def connect(self, close_with_result: bool = ...): ...  # type: ignore[override]
-    def invalidate(self, exception: Any | None = ...): ...
+    def invalidate(self, exception: Exception | None = ...) -> None: ...
     def detach(self) -> None: ...
-    def begin(self): ...
-    def begin_nested(self): ...
-    def begin_twophase(self, xid: Any | None = ...): ...
+    def begin(self) -> Transaction | None: ...
```

Comment:
Returning `None` from here seems to unlikely happen in practice. Maybe `Transaction | Any`? Also applies to at least `get_transaction()`.

Reply:
This indeed seems more like an implementation detail. I've just changed it to `Transaction` now.

`get_transaction()` on the other hand can be `None` legitimately, if `begin()` was not called.